{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Detailed description of software protocol\n",
    "## A fully reproducible one-stop-shop for the analysis of iTRAQ/TMT data\n",
    "\n",
    "This file serves as documentation for the IsoLabeled protocol including explanation of the folder structure, details on the available parameters, the output files and instruction how to access the underlying code.\n",
    "\n",
    "### Folders\n",
    "*IN*: This folder contains the files necessary to run the example, i.e. spectra file _iTRAQCancer.mgf_, human protein sequences (SwissProt) _sp_human.fasta_ and the experimental design _exp_design_example.tsv_. If you want to use this folder for your own analysis, you need to remove the spectra file.\n",
    "\n",
    "*OUT*: Fixed output folder for all intermediate and end result file. This folder is linked to the _data_ folder and therefore can be accessed from outside the Docker container.\n",
    "\n",
    "*data*: Starting IsoLabeledProtocol via the _run.sh_/_run.bat_ scripts automatically maps the current folder (on the host) to _data_. Placing spectra and sequence files into this folder crucially simplifies that analysis.\n",
    "\n",
    "*bin*: Here you can find the executables of [SearchGUI](http://compomics.github.io/projects/searchgui.html) and [PeptideShaker](http://compomics.github.io/projects/peptide-shaker.html).\n",
    "\n",
    "*Scripts*: Python scripts to handle parameter selection, execution of tools and R scripts\n",
    "\n",
    "### Parameters\n",
    "#### Folders and database search\n",
    "*Folder for spectra files (files need to be mgf) and fasta database:* \n",
    "We allow the folder _IN_ and _data_ where the latter is automatically mapped to the directory from which IsoLabeledProtocol is called (when using _run.sh_ or _run.bat_). The folder should contain the spectra files and a database of protein sequences given in a FASTA file. \n",
    "Raw files (e.g. Thermo raw files) can be converted into mgf using the msconvert tool of [ProteoWizard](http://proteowizard.sourceforge.net/tools.shtml). This requires a Windows computer. Files from different MS runs should be organized in folders as illustrated in Figure 1.\n",
    "\n",
    "\n",
    "![Illustration of folder structure](misc/ExperimentalDesigns.svg)\n",
    "**Figure 1**: Illustration of experimental designs and folder structure.\n",
    "\n",
    "\n",
    "\n",
    "#### Database search\n",
    "*Fasta file*: Select from FASTA files in the above given folder.\n",
    "The sequence headers of the file should be formatted in a [UniProt](http://uniprot.org)-like format. See more details on a suitable version of the FASTA format in the [SearchGUI documentation](https://github.com/compomics/searchgui/wiki/DatabaseHelp). Do not provide decoy sequences as they will be created automatically.\n",
    "\n",
    "*Precursor tolerance (ppm)*: MS tolerance in parts per million\n",
    "\n",
    "*Fragment ion tolerance (da)*: MSn tolerance in Dalton\n",
    "\n",
    "*Number of miscleavages*\n",
    "\n",
    "*Further fixed modifications*: Select Carbamidomethylation or None. iTRAQ and TMT labels as fixed modifications are added separately\n",
    "\n",
    "*Further variable modifications (Hold Ctrl to select multiple)*: Select variable modifications for the database search. Modified sequences will be used in the quantification and summarization of the proteins.\n",
    "\n",
    "*Target (protein, peptide, and PSM) FDR*: False discovery rate for the peptide identifications.\n",
    "\n",
    "*Quantification method*: Labelling method used in the experiment (iTRAQ/TMT)\n",
    "\n",
    "*Number of different conditions*: Number of different types of samples (e.g. time points, treatments, cell types). This parameter is needed to assign sample types to the different labeling channels.\n",
    "\n",
    "\n",
    "#### Experimental design\n",
    "This tab allows to  distribute the previously defined number of different experimental conditions to\n",
    " the channels of the MS runs. The default labels can be changed. The entire experimental design will be written into the table _exp_design.tsv_ after pressing the _Save design_ button. The file is located in th\n",
    "e _OUT_ folder.\n",
    "\n",
    "\n",
    "#### Quantitative analysis\n",
    "*Summarization method*: It is possible to select from the following methods to summarize peptide/PSM quantifications into protein abundance changes provided by the Bioconductor package [MSnbase](https://bioconductor.org/packages/release/bioc/html/MSnbase.html)\n",
    "\n",
    "- iPQF: feature-based weighting of peptide spectra according to this [paper](https://www.ncbi.nlm.nih.gov/pubmed/26589272)\n",
    "- Average over all PSMs (on log-scale)\n",
    "- Median over all PSMs (on log-scale)\n",
    "- Median over all PSMs after outlier removal (on log-scale)\n",
    "- Robust summarization using the R function `rlm` (on log-scale)\n",
    "\n",
    "\n",
    "*Minimum number of PSMs per protein*: Proteins or protein groups will only be quantified when 2 PSMs are available. This will be extended to unique peptides in the future\n",
    "\n",
    "*Use PTMs or quantification*: Include peptides decorated by variable PTMs like oxidations in the protein summarization.\n",
    "\n",
    "\n",
    "\n",
    "### Output files and figures\n",
    "\n",
    "All output files are available in the _OUT_ folder. For experiments with multiple runs, organised in subfolders, will report MS run specific results in corresponding subfolders of _OUT_ and general results in _OUT_.\n",
    "\n",
    "The database search via SearchGUI and validation via PeptideShaker result in a tab-delimited file (_experiment1_test_1_Extended_PSM_Report.txt_) containing detailed information from all peptide-spectrum matches (PSMs). Search and output parameters are automatically saved in json, _protocol_parameters.json_. Detailed information about identified mass spectra can be browsed importing the file _experiment.cpsx_ to your local version of PeptideShaker.\n",
    "\n",
    "Each spectra file is processed separately to match and quantify PSMs that passed the identification FDR (default 0.01). \n",
    "\n",
    "The mass distribution number of all matched fragment ions allows control for critical channels with inefficient labeling (*QC_ReporterIons_**filename**.pdf/png*). \n",
    "The output of all files of each run on the mass spectrometer are merged, normalized and visualized for quality control. Violin plots of median-normalized PSM intensities compare the intensity distributions (*QC_PSM_violinplots.pdf/png*). Channels with different distributions can identify problematic samples or changes within the entire proteome. All FDR-filtered and normalized PSM quantifications are saved to _AllQuantPSMs.csv_ and _AllQuantPSMs.RData_. \n",
    "\n",
    "\n",
    "Six different histograms counting PSM, peptide, protein and protein group numbers allows determining protein coverage and uniqueness by the available mass spectra (_QC_PSM_and_peptide_distribution.pdf/png_). \n",
    "\n",
    "Similarity between samples is assessed through scatter plots comparing all the quantifications from all ion channels (_QC_Pairwise_comparison.pdf/png_).\n",
    "\n",
    "Using the default parameters values, the PSMs are summarized to proteins by the iPQF summarization algorithm requiring a minimum of 1 PSM per protein, available through _AllQuantProteins.csv_. Other methods such as median summarization after outlier removal are available. \n",
    "\n",
    "A violin plot of protein ratios versus mean of all experimental conditions, available through _AllQuantProteinsInAllSamples.csv_, shows should exhibit similar distributions for the different channels (_QC_Protein_violinplots.pdf/png_). \n",
    "\n",
    "Quantifications from different runs (only one in the example) are merged and submitted to principal component analysis. With the scoring plot, one can look for similarity within replicates compared to different experimental conditions (_QC_Stat_PCA.pdf/png_). Completely mixed samples are unlikely to provide differentially regulated proteins.\n",
    "\n",
    "LIMMA statistical tests find regulated proteins.\n",
    "p-values and false discovery rates (p-values corrected for multiple testing) for pairwise comparisons (here X versus Y), all available through _DifferentiallyRegulatedProteins.csv_ , are visualized in histograms, volcano plots and a figure counting the number of differentially regulated proteins over a range of FDRs (*QC_Stat_Summary_**ConditionX-ConditionY**.pdf/png*). The latter can be used to identify a suitable combination of the confidence threshold and the number of significant proteins. It is advised to keep FDR<0.1 as the number of false positives becomes critically high otherwise.\n",
    "\n",
    "\n",
    "### Code structure and scripts\n",
    "#### Access\n",
    "\n",
    "All python and R scripts are easily accessible in the Docker image. Python scripts to run parameter seection and database search can be found in the folder _Scripts_. R scripts for downstream analysis can be made visibile by going to _View->Cell Toolbar->Hide code_ in the menu of the Example.ipynb Jupyter notebook. For the cells you want to see the code, uncheck the _Hide code_ checkbox. \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
